### PR TITLE
[TwigBridge][TwigBundle] add `asset_mapper_source()` function

### PR DESCRIPTION
**Pull request by kbond**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | n/a
| License       | MIT

This adds an `asset_mapper_source()` Twig function. The use-case is inlining css for emails:

```twig
{% apply inline_css(asset_mapper_source('vendor/foundation-emails/dist/foundation-emails.min.css'), asset_mapper_source('styles/email.css')) %}
    {# ... #}
{% endapply %}
```

This will also work with asset mapper assets that are compiled with css/less/tailwind/etc.

### CODE REVIEW

**Comment by kbond** [comment]
We do already have the `ImportMapExtension`/`Runtime` and I considered adding to it to avoid another extension. Also, wonder if should be called `importmap_source()`?

**Comment by smnandre** [comment]
> Also, wonder if should be called importmap_source()?

I don't think so as this is only related to asset mapper, which can be used outside any import map

File: src/Symfony/Bridge/Twig/Extension/AssetMapperExtension.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+final class AssetMapperExtension extends AbstractExtension
+{
+    public function getFunctions(): array
+    {
+        return [
+            new TwigFunction('asset_mapper_source', [AssetMapperRuntime::class, 'source'], ['is_safe' => ['all']]),
```
Comment (smnandre) [inline src/Symfony/Bridge/Twig/Extension/AssetMapperExtension.php:25]: I'm not sure about this `"is_safe" => ['all']` .. is this not a bit risky ?

Comment (kbond) [reply src/Symfony/Bridge/Twig/Extension/AssetMapperExtension.php:25]: I duplicated what `source()` from Twig core does so.. not sure.

File: src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            throw new \LogicException(\sprintf('The asset "%s" was not found.', $logicalPath));
+        }
+
+        if ($asset->content) {
+            return $asset->content;
+        }
+
+        if (false === $content = file_get_contents($asset->sourcePath)) {
+            throw new \RuntimeException(\sprintf('The asset "%s" could not be read.', $asset->sourcePath));
+        }
```
Comment (smnandre) [inline src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php:38]: I'm not sure here, or maybe the name is to reconsider. 

`$asset->content`, when provided, will always be different from `file_get_contents($asset->sourcePath)`

```php
    /**
     * The final content of this asset if different from the sourcePath.
     *
     * If null, the content should be read from the sourcePath.
     */
    public readonly ?string $content;
```

So i'm wondering if you should not always return the original content here 🤷‍♂️

Comment (kbond) [reply src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php:38]: Perhaps a different name than "source". This logic is definitely important for this feature (to be able to handle sass/tailwind/etc).

File: src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        if (!$asset = $this->assetMapper->getAsset($logicalPath)) {
+            throw new \LogicException(\sprintf('The asset "%s" was not found.', $logicalPath));
+        }
+
+        if ($asset->content) {
+            return $asset->content;
+        }
+
+        if (false === $content = file_get_contents($asset->sourcePath)) {
```
Comment (smnandre) [inline src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php:36]:
```suggestion
        if (false === $content = (new Filesystem())->readFile($asset->sourcePath)) {
```
?

As Filesystem is in AssetMapper dependencies..  and has now replaced all file_get_contents (afaik) ?

Comment (kbond) [reply src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php:36]: Had this originally but `Filesystem::readFile()` is 7.1+. Wasn't sure if we should bump deps.

File: src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php (hunk 2 above)
Comment (smnandre) [reply src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php:38]: I may have missunderstood... do you want the source code of the file... or the compiled one ?

Comment (kbond) [reply src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php:38]: Compiled if available or the source of not.

My example in the PR description: you can imagine foundation CSS being just the source and style.css being complied via tailwind.

Comment (kbond) [reply src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php:38]: For some context, I'm trying to do what Ryan is showing here:

https://symfonycasts.com/screencast/mailer/encore-css
https://symfonycasts.com/screencast/mailer/encore-inline_css

But with asset-mapper.

Comment (smnandre) [reply src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php:38]: This was what made me question the name.. So in the end this is the content of the mapped asset more than its source code.

Maybe `mapped_asset_content` ? `asset_mapper_compiled_content` ?

Comment (kbond) [reply src/Symfony/Bridge/Twig/Extension/AssetMapperRuntime.php:38]: I see, you're right, content makes more sense.

**Comment by kbond** [comment]
Closing for now as @smnandre pointed out, using in production could be trouble!

**Pull request closed without merging**